clippy_lints/src/strings.rs

```diff
@@ -451,3 +451,49 @@ impl<'tcx> LateLintPass<'tcx> for StringToString {
         }
     }
 }
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Warns about calling `str::trim` (or variants) before `str::split_whitespace`.
+    ///
+    /// ### Why is this bad?
+    /// `split_whitespace` already ignores leading and trailing whitespace.
+    ///
+    /// ### Example
+    /// ```rust
+    /// " A B C ".trim().split_whitespace()
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// " A B C ".split_whitespace()
+    /// ```
+    #[clippy::version = "1.61.0"]
+    pub TRIM_SPLIT_WHITESPACE,
+    pedantic,
+    "using `str::trim()` or alike before `str::split_whitespace`"
+}
+declare_lint_pass!(TrimSplitWhitespace => [TRIM_SPLIT_WHITESPACE]);
+
+impl<'tcx> LateLintPass<'tcx> for TrimSplitWhitespace {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &Expr<'_>) {
+        if_chain! {
+            if let ExprKind::MethodCall(path, [recv], split_ws_span) = expr.kind;
+            if path.ident.name == sym!(split_whitespace);
+            if let ExprKind::MethodCall(path, [recv], trim_span) = recv.kind;
+            if let trim_fn_name @ ("trim" | "trim_start" | "trim_end") = path.ident.name.as_str();
+            let recv_ty = cx.typeck_results().expr_ty(recv).peel_refs();
+            if recv_ty.is_str() || is_string(cx, recv);
```

Comment:
This only checks for `&str` and `String`. So anything implementing `Deref<Target=str>` will not be linted. Do we want that?

Reply:
Yes. It should run if those methods are chained regardless of whether it's through a deref.

Reply:
I don't think this is what clippy usually does. [In this example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3b103af0ca46aaf5556113e9a8a7a402) only the `&str` version is triggering a `str_to_string` lint. The version using a wrapping struct implementing `Deref<Target=str>` does not.

Reply:
The `str_to_string` lint only warns on `&str`s because it is specially about calling `to_string` on `str` -  [str_to_string docs](https://rust-lang.github.io/rust-clippy/master/index.html#str_to_string) 

This lint is applicable to any type that `.trim().split_whitespace()` can be used on. You just need to be careful with your suggestion if the type has it's own `split_whitespace` method.